src/handler/sign_psbt.c

```diff
@@ -649,8 +650,8 @@ init_global_state(dispatcher_context_t *dc, sign_psbt_state_t *st) {
     // Swap feature: check that wallet policy is a default one
     if (G_swap_state.called_from_swap && !st->is_wallet_default) {
         PRINTF("Must be a default wallet policy for swap feature\n");
-        SEND_SW(dc, SW_INCORRECT_DATA);
-        return false;
+        SEND_SW(dc, SW_FAIL_SWAP);
+        finalize_exchange_sign_transaction(false);
```

Comment:
An issue that was indeed already present (and therefore, we could solve in a separate PR) is that `finalize_exchange_sign_transaction` is only called if there is an error in the logic strictly related to swaps, but there are other places where an error could happen.

The right way of fixing this might be to remove all the `SEND_SW` in the code, and instead use the return values to signal/propagate errors.

We could take the change to do a bigger refactoring error reporting more generally, for example with a short error message in the response APDU instead of just a status word. That would help debugging, as people having issues only know that they are getting "INCORRECT_DATA", but there are tens of places where that could be returned...

Reply:
+1